### ISSUE TITLE
Support mosaic optimizations as plugins

### ISSUE THREAD

**Issue by williamFalcon** [PyTorchLightning/pytorch-lightning#12360]
This library [mosaic](https://github.com/mosaicml/composer) has neat tricks for optimizing the models for faster training. 
Each application is done as a single line to the model
```python
import composer.functional as cf
from torchvision import models

my_model = models.resnet18()

# add blurpool and squeeze excite layers
model = cf.apply_blurpool(my_model)
model = cf.apply_squeeze_excite(my_model)

# your own training code starts here
```

Which is something we can automatically do for users under the hood if they want to enable the mosaic optimizations.

cc @Borda @carmocca @tchaton

**Comment by carmocca** [comment]
I believe the proposed API just mimics what composer offers:

```python
trainer = composer.Trainer(
    ...
    algorithms=[
        BlurPool(replace_convs=True, replace_maxpools=True, blur_first=True),
        ChannelsLast(),
        CutMix(num_classes=10),
        LabelSmoothing(smoothing=0.1),
    ]
)
```
(from their README)

The key difference here is that (AFAIK) `composer` does not provide a "Module" abstraction such as the `LightningModule` so it is natural for their library to put these directly in their trainer.

Also, looking at the source code, it does not look like they do any special management of these and just run all algorithms during every trigger event:

https://github.com/mosaicml/composer/blob/42271f8f6b10810d660318d17d037822beb05ee7/composer/core/engine.py#L177-L185
https://github.com/mosaicml/composer/blob/42271f8f6b10810d660318d17d037822beb05ee7/composer/core/engine.py#L192-L196

These could be directly part of the `LightningModule` especially because they are part of the research and that's the where you usually put your research code. So unless we need to integrate with strategies or something specific to our internals, I agree with @ananthsub

**Comment by carmocca** [comment]
@hanlint do you forsee any limitations that would be better adressed by making the algorithms part of the `Trainer`?

**Comment by hanlint** [comment]
@carmocca agreed with the approach. We designed the functional API so that users can utilize our methods inside their own training loops, so would be natural for users wishing to employ our efficiency methods to put in `LightningModule` themselves, where the research code lives.

A few limitations I can think of:
* The composer trainer does handle some of the ordering complexity e.g. the algorithms run in LIFO during the `before_` and `after_` events so that they clean up their own effects. Or ensuring that `SelectiveBackprop` runs first (see: https://github.com/mosaicml/composer/blob/0d7175573225001549e36ac91b2e3def250fa19f/composer/core/engine.py#L231). 
* Less user error in inserting the algorithms at the wrong place in the loop.
* Other quality of life such as tracking when algorithms are run, algorithm-specific profiling overlays, serialization/resumption of algorithms state, etc.

However, most of these are ease-of-use items that could be handled with good docs and warnings (e.g. we recently added algorithm warnings in https://github.com/mosaicml/composer/pull/720), as we harden our functional API.

**Comment by tchaton** [comment]
@[hanlint](https://github.com/hanlint) I wanted to introduce you to Lightning Flash: https://github.com/PyTorchLightning/lightning-flash/tree/master/flash.

We have 2 custom objects there: [Input and InputTransform](https://github.com/PyTorchLightning/lightning-flash/blob/master/flash_examples/flash_components/custom_data_loading.py) used to organize data loading and data transform + the concept of Adapter ([learn2learn](https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/image/classification/adapters.py#L74) example) to easily integrate thrid party libraries.
I believe we could explore a Flash integration first, understand how we can address some of the limitations raised above, and upstream any core components which work for all users.

**Comment by hanlint** [comment]
Thanks @tchaton for the pointer, I will take a look.
